docs/interfaces.md

```diff
@@ -0,0 +1,431 @@
+        Evidence []byte,        // attestation evidence (epid/dcap/etc...)
+}
+
+type CCKeyRegistrationMessage struct {
+        Chaincode_EK []byte,    // public chaincode encryption key
+        Enclave_VK []byte,      // creator of this message
+        Sig_Enclave []byte,     // signature of the message creator
+}
+
+type ExportMessage struct {
+        Chaincode_EK []byte,    // public chaincode encryption key
+        CCKeys_enc []byte,      // chaincode keys encrypted for the receiver
+        Enclave2_VK []byte,     // creator of this message
+        Enclave_VK []byte,      // receiver of this message
+        Sig_Enclave []byte,     // signature of the message creator
+}
+```
+
+ERCC keeps in instance of an attestation.Verifier to verify the attestation evidence. In case of EPID, the evidence is an `IASAttestationReport` object. This instance has the Intel attestation verification key.
+```go
+type EnclaveRegistryCC struct {
+	ra  attestation.Verifier
+}
+
+// Intel IAS verification key (same as used by TLCC Enclave)
+const IntelPubPEM = `...`
+
+type IASAttestationReport struct {
+	EnclavePk                   []byte `json:"EnclavePk"`
+	IASReportSignature          string `json:"IASReport-Signature"`
+	IASReportSigningCertificate string `json:"IASReport-Signing-Certificate"`
+	IASReportBody               []byte `json:"IASResponseBody"`
+}
+```
+
+
+# TLCC
+
+## Interface:
+The TLCC interface is implemented through a "normal" chaincode interface.
+
+```go
+// chaincode interface 
+func Invoke(stub shim.ChaincodeStubInterface) pb.Response {}
+
+// TODO note that we are multiplexing normal and range queries with this method
+// used by ecc via cc2cc
+getMetadata(key string, nonce string, isRangeQuery bool) (cmacBase64  string, error)
+
+validateMSP(mspId string, nonce string) (valid bool, cmacBase64 string, error)
+
+// used by the admin to join the channel
+joinChannel(channel_id string) error
+```
+
+```c++
+// interface exposed to TLCC_enclave via cgo
+// logging
+void ocall_print_string([in, string] const char *str);
+```
+
+## State:
+TLCC keeps a local reference of the enclave id. This reference is volatile. Also a reference to the peer is kept to access the ledger and read blocks.
+
+```go
+type TrustedLedgerCC struct {
+	enclave enclave.Stub
+	peer    *peer.Peer
+}
+
+type StubImpl struct {
+	eid C.enclave_id_t
+}
+```
+
+
+# TLCC_Enclave
+
+## Interface:
+Defined in the TLCC EDL.
+```c++
+// TBD is this interface already sufficient to support multi-channel?
+// TODO check if a block contatains any channel information?
+
+// creates a new channel
+public int ecall_join_channel(
+        [in, size=gen_len] uint8_t *genesis, uint32_t gen_len);
+
+// pushes next block to tlcc
+public int ecall_next_block(
+        [user_check] uint8_t *block_bytes, uint32_t block_size);
+
+// verify state data
+public int ecall_get_state_metadata(
+        [in, string] const char *key, // TBD: should we add namespace separetely?
```

Comment:
I think adding the namespace is useful. Certainly would allow ecc to access ercc state meta-data and might safe us a separate function which we currently have in the UML to validate ercc registration state.  And as mentioned above, might be good to already make the channel also explicit?  

BTW: in join the channel-id is implicitly the genesis block but in other cases/externally a channel-id is just a string: how do we do that mapping and is there any issue of having the string or would we need a cryptographic secure identifier (e.g., hash of genesis block)?  My intuition is that name probably should be ok (or fabric already would have some problem), in which case maybe the easiest is to just add a name parameter to the `join_channel`? [PS: seeing only now your `TODO` above, maybe this mapping channel data to channel name was maybe exactly the motivation for that?]

Reply:
namespaces added. I also added a note saying that we have a tlcc_enclave for each channel, that is, there is no need to make the channel_id explicit when passing a block to tlcc_enclave. the `join_channel` provided by `tlcc` indeed requires the chaincode_id as known by the peer to start the block listener. 

Maybe tlcc_enclave should get a method to query the root-of-rust, that, is the cryptographic hash of the genesis block it has received from the pee. Ideally this statement is attested?

Reply:
if we pass _only_ channel name, how do we ensure though that it really matches the genesis block seen by tlcc_enclave?  Couldn't a malicious org hosting the designated peer create a shadow ledger for tlcc which mimics everything on the real except being rooted in bogus orderer/msp root one and where the malicous org could inject bogus requests to reveal data when useful? somewhere/somehow we have to make sure  that untrusted and trusted ledger share the same genesis block.  One way could be that the gensis block is attested by TLCC when authenticating to ecc, ecc has genesis block baked in & attested to similar to tlcc-mrenclave which then both would be verified by ercc during registration?